okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -143,130 +273,348 @@ private static String checkOneDependency(ModuleDescriptor md, InterfaceDescripto
-   */
-  public static String checkDependencies(Collection<ModuleDescriptor> available,
-                                         Collection<ModuleDescriptor> testList,
-                                         boolean removeIfMissingDep) {
-    Map<String, List<InterfaceDescriptor>> ints = getProvidedInterfaces(available);
-    List<String> list = new LinkedList<>();
+  static class ModuleInterface {
+    final ModuleDescriptor moduleDescriptor;
+    final InterfaceDescriptor interfaceDescriptor;
 
-    Iterator<ModuleDescriptor> iterator = testList.iterator();
-    while (iterator.hasNext()) {
-      ModuleDescriptor md = iterator.next();
-      List<String> res = checkDependenciesInts(md, available, ints);
-      if (!res.isEmpty()) {
-        if (removeIfMissingDep) {
-          available.remove(md);
-          ints = getProvidedInterfaces(available);
-          iterator.remove();
-          iterator = testList.iterator();
-        } else {
-          list.addAll(res);
+    ModuleInterface(ModuleDescriptor md, InterfaceDescriptor id) {
+      this.moduleDescriptor = md;
+      this.interfaceDescriptor = id;
+    }
+  }
+
+
+  private static void sortTenantModules(List<TenantModuleDescriptor> tml,
+      Map<String, ModuleDescriptor> modsAvailable, Map<String, ModuleDescriptor> modsEnabled) {
+
+    Set<String> added = new HashSet<>();
+    // make a list of all modules involved.. also those removed.
+    LinkedList<ModuleDescriptor> sortedList = new LinkedList<>();
+    for (TenantModuleDescriptor tm: tml) {
+      if (tm.getAction() == TenantModuleDescriptor.Action.enable
+          || tm.getAction() == TenantModuleDescriptor.Action.uptodate
+          || tm.getAction() == TenantModuleDescriptor.Action.disable) {
+        sortedList.add(modsAvailable.get(tm.getId()));
+        added.add(tm.getId());
+      }
+    }
+    for (ModuleDescriptor md : modsEnabled.values()) {
+      if (!added.contains(md.getId())) {
+        sortedList.add(md);
+        added.add(md.getId());
+      }
+    }
+    topoSort(sortedList);
+    // we now have a list where things mentioned in the install comes first.. Thus, if
+    // for cases where there are different orders satisfying dependencies, the order in the
+    // install is honored and will be listed first.
+    logger.info("Topo sort result {}", () ->
+        sortedList.stream()
+        .map(ModuleDescriptor::getId)
+            .collect(Collectors.joining(", ")));
+
+    logger.info("Input install list {}", () ->
+        tml.stream()
+            .map(TenantModuleDescriptor::getId)
+            .collect(Collectors.joining(", ")));
+
+    List<TenantModuleDescriptor> tml2 = new ArrayList<>();
+
+    // go through disabled modules
+    Iterator<ModuleDescriptor> moduleIterator = sortedList.descendingIterator();
+    while (moduleIterator.hasNext()) {
+      ModuleDescriptor md = moduleIterator.next();
+      Iterator<TenantModuleDescriptor> iterator = tml.iterator();
+      while (iterator.hasNext()) {
```

Comment:
This nested loop could be replaced by a hash lookup. Simply set up a `Set<String> disable = new HashSet<>()` similar to the existing added HashSet.
The same below for enable.

Reply:
Could be done.. I tried, but not so easy.. The same module could be mentioned twice.. That could be the case for enable, uptodate actions..